net/dns/resolver/forwarder.go

```diff
@@ -164,6 +165,23 @@ func clampEDNSSize(packet []byte, maxSize uint16) {
 	binary.BigEndian.PutUint16(opt[3:5], maxSize)
 }
 
+// dnsForwarderFailing should be raised when the forwarder is unable to reach the
+// upstream resolvers. This is a high severity warning as it results in "no internet".
+// This warning must be cleared when the forwarder is working again.
+//
+// We allow for 5 second grace period to ensure this is not raised for spurious errors
+// under the assumption that DNS queries are relatively frequent and a subsequent
+// successful query will clear any one-off errors.
+var dnsForwarderFailing = health.Register(&health.Warnable{
+	Code:                "dns-forward-failing",
+	Title:               "DNS unavailable",
+	Severity:            health.SeverityHigh,
+	DependsOn:           []*health.Warnable{health.NetworkStatusWarnable},
+	Text:                health.StaticMessage("Tailscale can't reach the configured DNS servers. Internet connectivity may be affected."),
```

Comment:
I think it'd be good to have this be more specific, and say that there were no upstream servers configured in the OS, and none set in the tailnet override, perhaps pointing to a KB article, when there are no servers.

When there are resolution timeouts, we should report the number of failed DNS queries in the last N minutes.

Reply:
Ack.. I've added the resolvers to the args so we can disambiguate betwen "no resolvers" and "all resolvers failed".  There's a strategy in place to map the codes to KB articles for all human-visible warnings but we're not quite there yet...